File: RESTful Samples/Ruby/Test/TestClient.rb
require 'test/unit/ui/console/testrunner'
require 'test/unit'
require 'uri'
require_relative '..\report'
require_relative '..\reportpdf'
require_relative '..\XmlDataSource'
require_relative '..\AdoDataSource'
require_relative '..\TemplateVariable'
require_relative '..\Report'

class TestClient < Test::Unit::TestCase

	@uri
	def setup()
		@uri = URI("http://localhost:8080")
	end
	def test_GetVersion()
		v = Report::GetVersion @uri
		assert(v != nil)
		assert(v.EngineVersion.is_a?(String))
		assert(v.ServiceVersion.is_a?(String))
	end
	
	def test_PostTemplateReturnsReportPdf
		output = StringIO.new()
		template = File.open('TestFiles\Sample1.docx', 'rb')
		report = ReportPdf.new(@uri, template, output)
		report.Process()
		assert(output.size > 8)
		bytes = Array.new
		bytes = output.string[0, 8]
		expected = ['25', '50', '44', '46', '2d', '31', '2e', '35']
		for i in 0..7
			assert((bytes[i]).unpack('H*')[0] == expected[i])
		end
	end
	
	def test_PostTemplateWithXmlData()
		xmlFile = File.open('TestFiles\Manufacturing.xml', 'rb')
		dataSources = {"MANF_DATA_2009" => XmlDataSource.new(xmlFile)}
		templateFilePath = 'TestFiles\Manufacturing.docx';
		outputFilePath = 'TestFiles\XmlDataOutput.pdf';
		templateFile = File.open(templateFilePath, 'rb')
		outputFile = File.new(outputFilePath, 'wb')
		report = ReportPdf.new(@uri, templateFile, outputFile)
		report.ProcessDataSources(dataSources)
	
	end
	
	def test_PostTemplateWithAdoData()
		dataSources = {"MSSQL" => AdoDataSource.new("System.Data.SqlClient", "Data Source=mssql.windward.net;Initial Catalog=Northwind;User=demo;Password=demo")}
		templateFilePath = 'TestFiles\MsSqlTemplate.docx';
        outputFilePath = 'TestFiles\AdoDataOutput.pdf';
		templateFile = File.open(templateFilePath, 'rb')
		outputFile = File.new(outputFilePath, 'wb')
		report = ReportPdf.new(@uri, templateFile, outputFile)
		report.ProcessDataSources(dataSources)
	end
	
	def test_ClientVariables()
		ds = XmlDataSource.new(File.open('TestFiles\\Manufacturing.xml', 'rb'))
		templateVars = [TemplateVariable.new("var1", "Hi there")]
		ds.variables = templateVars
		dataSources = {"MANF_DATA_2009" => ds}
		templateFile = File.open('TestFiles\Variables.docx', 'r')
		outputFile = File.new('TestFiles\VariablesOutput.pdf', 'w')
		report = ReportPdf.new(@uri, templateFile, outputFile)
		report.ProcessDataSources(dataSources)
	end
	
	def test_PostTemplateAsync()
		xmlFile = File.open('TestFiles\\Manufacturing.xml')
		dataSources = {"MANF_DATA_2009" => XmlDataSource.new(xmlFile)}
		templateFilePath = 'TestFiles\Manufacturing.docx';
		outputFilePath = 'TestFiles\AsyncOutput.pdf';
		templateFile = File.open(templateFilePath, 'rb')
		report = ReportPdf.new(@uri, templateFile)
		report.ProcessDataSources(dataSources)
		
		while(report.GetStatus == :Working)
			sleep(100)
		end
		
		if(report.GetStatus() == :Ready)
			File.new(outputFilePath, 'wb') do |output|
				output.write(report.GetReport())
			end
		end
	end
end
Test::Unit::UI::Console::TestRunner.run(TestClient)

File: RESTful Samples/RESTfulEngine.TypeScriptClient/TypeScriptClient/tests/TestClient.ts
import assert = require('assert');
import client = require("../src/Client")
import version = require("../src/Version")
import Fs = require("fs");
import Html = require("../src/report/ReportHtml");
import ReportHtml = Html.ReportHtml;
import Ado = require("../src/datasource/AdoDataSource");
import AdoDataSource = Ado.AdoDataSource;
import Xml = require("../src/datasource/XmlDataSource");
import XmlDataSource = Xml.XmlDataSource;
import variable = require("../src/datasource/TemplateVariable");
import TemplateVariable = variable.TemplateVariable;
import Json = require("../src/datasource/JsonDataSource");
import JsonDataSource = Json.JsonDataSource;
import Utils = require("../Utils");

var testClient = new client.Client("localhost", 8080);
var TEST_FILES_DIRECTORY = "../files/";

export function TestGetVersionRequest() {
    testClient.getVersion((version: version.Version) => {
        assert.ok(version.majorEngineVersion() > 13);
    });
}

export function TestRunReportNoDataSource() {
    var report = new ReportHtml(Utils.readFile(TEST_FILES_DIRECTORY + "test-no-datasource.docx"));

    testClient.runReport(report, (response: string) => {
        assert.ok(response.search("Test") > 0);
    });
}

export function TestReportAdoDataSource() {
    var report = new ReportHtml(Utils.readFile(TEST_FILES_DIRECTORY + "test-ado-datasource.docx"));
    var data = new AdoDataSource("System.Data.SqlClient", "Data Source=mssql.windward.net;Initial Catalog=Northwind;User=demo;Password=demo");
    report.Datasources["MSSQL"] = data;

    testClient.runReport(report, (response: string) => {
        assert.ok(response.search("Davolio") > 0);
    });
}

export function TestRunReportXmlDataSource() {
    var report = new ReportHtml(Utils.readFile(TEST_FILES_DIRECTORY + "test-xml-datasource.docx"));
    var data = new XmlDataSource(Utils.readFile(TEST_FILES_DIRECTORY + "xml-test.xml"));
    report.Datasources["WRTRUCK1"] = data;

    testClient.runReport(report, (response: string) => {
        assert.ok(response.search("Windward Dashboard Summary Report") > 0, response);
    });
}

export function TestRunReportJsonDataSource() {
    var report = new ReportHtml(Utils.readFile(TEST_FILES_DIRECTORY + "test-json-datasource.docx"));
    var data = new JsonDataSource("http://json.windward.net/Northwind.json");
    report.Datasources[""] = data;

    testClient.runReport(report, (response: string) => {
        assert.ok(response.search("Davolio") > 0, response);
    });
}

export function TestRunReportWithTemplateVariables() {
    var report = new ReportHtml(Utils.readFile(TEST_FILES_DIRECTORY + "test-template-variables.docx"));
    var data = new AdoDataSource("System.Data.SqlClient", "Data Source=mssql.windward.net;Initial Catalog=Northwind;User=demo;Password=demo");
    data.Variables = [new TemplateVariable("order", "10538")];

    report.Datasources["MSSQL"] = data;

    testClient.runReport(report, (response: string) => {
        assert.ok(response.search("10538") > 0);
    });
}

export function TestRunReportAsync() {
    var report = new ReportHtml(Utils.readFile(TEST_FILES_DIRECTORY + "test-no-datasource.docx"));

    testClient.runReportAsync(report, (guid: string) => {
        assert.ok(guid.length === 36);
    });
}

export function TestRunReportAsyncAndGetReport() {
    var report = new ReportHtml(Utils.readFile(TEST_FILES_DIRECTORY + "test-no-datasource.docx"));

    testClient.runReportAsync(report, (guid: string) => {
        assert.ok(guid.length === 36);
        sleep(5);

        testClient.getReport(guid, (response) => {
            assert.ok(response.search("Test") > 0);
        });
    });
}

// don't use this in a production app
function sleep(seconds) {
    var e = new Date().getTime() + (seconds * 1000);
    while (new Date().getTime() <= e) { }
}